File: precision.f90
!######################################!
!                                      !
!     Precision Test                   !
!                                      !
!     FrB                              !
!                                      !
!######################################!

!
! Description
! ===========
! Module containing precision info subroutines
!
!

module precision_module

    use types

    implicit none

    real(sp) :: single_precision_mi
    real(dp) :: double_precision_mi

contains

    subroutine print_precision(single_precision_in, double_precision_in)

        implicit none

        real(sp), intent(in) :: single_precision_in
        real(dp), intent(in) :: double_precision_in

        print*,
        print*, ">>> From Fortran: Precision Prints:"
        print*,

        print*, "Module internal: ISO single precision = ", single_precision_mi
        print*, "    Precision: ", precision(single_precision_mi)
        print*, "    Range    : ", range(single_precision_mi)
        print*, "    Tiny     : ", tiny(single_precision_mi)
        print*, "    Huge     : ", huge(single_precision_mi)
        print*, "    Digits   : ", digits(single_precision_mi)
        print*, "    Max exp  : ", maxexponent(single_precision_mi)
        print*, "    Min exp  : ", minexponent(single_precision_mi)
        print*, "    Kind     : ", kind(single_precision_mi)
        print*,

        print*, "Module internal: ISO double precision = ", double_precision_mi
        print*, "    Precision: ", precision(double_precision_mi)
        print*, "    Range    : ", range(double_precision_mi)
        print*, "    Tiny     : ", tiny(double_precision_mi)
        print*, "    Huge     : ", huge(double_precision_mi)
        print*, "    Digits   : ", digits(double_precision_mi)
        print*, "    Max exp  : ", maxexponent(double_precision_mi)
        print*, "    Min exp  : ", minexponent(double_precision_mi)
        print*, "    Kind     : ", kind(double_precision_mi)
        print*,

        print*, "Intent in: ISO single precision = ", single_precision_in
        print*, "    Precision: ", precision(single_precision_in)
        print*, "    Range    : ", range(single_precision_in)
        print*, "    Tiny     : ", tiny(single_precision_in)
        print*, "    Huge     : ", huge(single_precision_in)
        print*, "    Digits   : ", digits(single_precision_in)
        print*, "    Max exp  : ", maxexponent(single_precision_in)
        print*, "    Min exp  : ", minexponent(single_precision_in)
        print*, "    Kind     : ", kind(single_precision_in)
        print*,

        print*, "Intent in: ISO double precision = ", double_precision_in
        print*, "    Precision: ", precision(double_precision_in)
        print*, "    Range    : ", range(double_precision_in)
        print*, "    Tiny     : ", tiny(double_precision_in)
        print*, "    Huge     : ", huge(double_precision_in)
        print*, "    Digits   : ", digits(double_precision_in)
        print*, "    Max exp  : ", maxexponent(double_precision_in)
        print*, "    Min exp  : ", minexponent(double_precision_in)
        print*, "    Kind     : ", kind(double_precision_in)
        print*,

    end subroutine print_precision


end module precision_module

!######################################!
!                                      !
!     Precision Test                   !
!                                      !
!     FrB                              !
!                                      !
!######################################!

File: precision.py
#! /usr/bin/env python

########################################
#                                      #
#     Precision Test                   #
#                                      #
#     FrB                              #
#                                      #
########################################

"""
Description
===========
Python script to test import of fortran module and
precision of passed arguements

"""

import numpy as np
import precision

single_precision_in = np.float32(1.0)
double_precision_in = np.float64(1.0)

# arrays:
d32 = np.dtype(np.float32)
d64 = np.dtype(np.float64)
np_array32 = np.array([32, 10, 3, 2, 1], dtype = d32)
np_array64 = np.array([1, 2, 3, 10, 64], dtype = d64)


print " >>> Python declarations:"
print
print "Numpy: single precision = {0}".format(single_precision_in)
print np.finfo(single_precision_in)
print
print "Numpy: double precision = {0}".format(double_precision_in)
print np.finfo(double_precision_in)
print
print "Numpy: single precision array = {0}".format(np_array32)
print np.finfo(np_array32[0])
print
print "Numpy: double precision array = {0}".format(np_array64)
print np.finfo(np_array64[0])
print

# through interface
precision.py_interface(single_precision_in, double_precision_in, np_array32, np_array64)

# through module/subroutine
#precision.precision_module.print_precision(single_precision_in, double_precision_in)

########################################
#                                      #
#     Precision Test                   #
#                                      #
#     FrB                              #
#                                      #
########################################
